### DISCUSSION TITLE
Best way to do merge conflict

### DISCUSSION THREAD

**Discussion by RyanQuey** [abraunegg/onedrive#2293] [Q&A] [answered]
Is there a good way to automate merge conflicts? Currently I see the two conflict files in my directory when it happens, which is good, and I can fix it manually (e.g., using [meld](https://meldmerge.org/)), but is there an established easy way to do this? 

Just trying to avoid writing my own bash script for this haha :)

Thanks

**Comment by abraunegg** [comment, marked as answer]
@RyanQuey 
> Is there a good way to automate merge conflicts? Currently I see the two conflict files in my directory when it happens, which is good, 

By conflict files I assume here you are talking about files that have your local hostname appended to them?

Getting these conflicts should actually be fairly rare - however, if you are using older client versions (which are obsolete and unsupported) then the possibility of getting them increases dramatically. Please ensure you are using client version [![Version](https://img.shields.io/github/v/release/abraunegg/onedrive)](https://github.com/abraunegg/onedrive/releases) or greater. Check with `onedrive --version` and upgrade your client if required.

The issue also occurs due to SharePoint and how it physically changes files post upload, thus your local copy is never the same and can contribute to creating the issue. To help overcome this problem with SharePoint the application also has the following configuration setting in the config file only:
```
bypass_data_preservation = "false"
```

Set this to "true" in your configuration file and restart your client. to avoid the SharePoint 'feature' from causing the issue.

> and I can fix it manually (e.g., using [meld](https://meldmerge.org/)), but is there an established easy way to do this?

There is no established way to fix your conflict files. That is your data and I have zero insight into.

A fix for me would be to actually understand what is causing them in the first place, and fixing that.

**Comment by RyanQuey** [comment]
I only just started using this library, so maybe I'm using it wrong, but it
is the latest version. I'm not using SharePoint either. But my vault has
over 10,000 files and 18 GB of files,  not tons but it's taking about 10
minutes to fully sync all the files after making changes on a remote
machine.

I tested this again, it seems to be happening most when it says that it has
to start from scratch and erase all state, but even just for a normal sync,
a couple sync conflicts are coming up.

I will keep experimenting though to try to see why I'm getting merge
conflicts. To clarify: you're saying that I really shouldn't be getting any
conflicts at all, is that correct?

**Reply by RyanQuey** [reply]
fwiw, I came across this thread, which is for syncthing but can be adjusted for use with this as well: 

- Script: https://github.com/joopdo/dotfiles/blob/master/scripts/syncconflict.sh
- Discussion: https://github.com/syncthing/syncthing/issues/1746

I adjusted for Onedrive Linux here (not perfect but worked enough to be helpful for my use case):
- [For using `meld` to merge each file](https://gist.github.com/RyanQuey/727a3285b7716fe1311cfb752d406f0b)
- [For just listing out each file](https://gist.github.com/RyanQuey/27ccca8d2c3b7d6a005bfaad5dfb1974)

**Comment by abraunegg** [comment]
@RyanQuey 
> I only just started using this library, so maybe I'm using it wrong, but it is the latest version.

Please can you confirm this via `onedrive --version` ?

> I tested this again, it seems to be happening most when it says that it has to start from scratch and erase all state, but even just for a normal sync, a couple sync conflicts are coming up.

This is *very* odd for a number of reasons:

1. Why is the application saying it has to start from scratch all the time? That in itself is odd behavior. Are you *constantly* changing your application configuration??? Nothing thus far indicates this.
2. That conflicts are occurring during a normal sync - indicates something is up.

So I am going to make some broad assumptions here:

1. You probably are running Ubuntu
2. You probably installed from Ubuntu Universe or the Ubuntu PPA - meaning you are hitting this issue: https://github.com/abraunegg/onedrive/issues/1274
3. You then maybe installed from the OpenSuSE repository - but did not remove the offending systemd entry and/or still have the older version running on your system

This combination would explain why files are coming up as conflicts all the time and why the application may be saying it has to start over due to DB corruption or something along those lines ....


So ... Please confirm what application version you are running and *how* you installed it & on what OS you are running the client on.

In addition to this - please start providing concrete evidence of what you are seeing and experiencing by providing verbose logs and debug logs as per https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support

**Reply by RyanQuey** [reply]
Sorry, I wasn't clear above - I did check the version using `onedrive --version` and it returned: 
```
onedrive v2.4.23-1+np1
```

I am running Ubuntu, and I did install using OpenSuSE repository

It is not forcing a refresh every time - it did happen one time though, after I changed some configs, then without doing the "full resync" shutdown my machine. The next day when I booted back up, there were quite a few conflicts, but since then the problem has been much more minimal. Just now when I ran a sync that changed perhaps upward of 200 files there were no problems at all. 

I think we can close this thread, your answer above was helpful - on the whole the one time there were lots of conflicts was because I didn't run a "full resync" in a timely fashion as mentioned in the docs, and from here on out I should not have merge conflicts like that come up anymore, as long as I keep the lib up to date :) 

Thanks again
